### DISCUSSION TITLE
Integrating other language scripts with bpftrace

### DISCUSSION THREAD

**Discussion by techneophyte** [bpftrace/bpftrace#3234] [Q&A]
I am new to bpftrace. Can someone help me with examples on how can I add C/C++ functions inside a .bt file which I can call in a tracepoint syscall? Or, examples of tracing events in C++ script using bpftrace? I could not find any examples online

**Comment by jordalgo** [comment]
Hi @techneophyte, thanks for your question.

> Can someone help me with examples on how can I add C/C++ functions inside a .bt file 

This is not possible at the moment but being able to interop with C code/functions is on our list of features we want to support soon.

> Or, examples of tracing events in C++ script using bpftrace?

So this kinda depends on what you want to do. Right now you can trace user space functions (with bpftrace or BPF in general) via uprobes: https://github.com/bpftrace/bpftrace/blob/master/man/adoc/bpftrace.adoc#probes-uprobe

And if you specify the cpp prefix, bpftrace will do the demangling for you. Example of tracing bpftrace itself:
```
bpftrace:cpp:"bpftrace::BPFtrace::add_probe" { ... }
```

There are also USDTs but you'd need to add those tracepoints into the code yourself. 

For tracing syscalls, you can do something like this:
```
BEGIN {
	printf("Tracing open syscalls... Hit Ctrl-C to end.\n");
}

tracepoint:syscalls:sys_enter_open,
tracepoint:syscalls:sys_enter_openat {
	printf("%-6d %-16s %s\n", pid, comm, str(args.filename));
}
```

Most of these examples, I took from [the adoc](https://github.com/bpftrace/bpftrace/blob/master/man/adoc/bpftrace.adoc)

**Comment by techneophyte** [comment]
@jordalgo Thanks for your response. I have a quick question. I am using [killsnoop.bt](https://github.com/bpftrace/bpftrace/blob/master/tools/killsnoop.bt) for my usecase. As part of the `sys_enter_kill` tracepoint, I want to retrieve the information of the user who initiated this request from `/proc/pid/environ`. I want to save a copy this file to a different location as this might be deleted when the process ends. Can you provide any suggestions on how can I do this?

**Reply by ajor** [reply]
I think it'd be better to try to use the bpftrafce builtin variables `uid` and `gid` to find out the user who invoked `sys_enter_kill`. That way everything stays inside BPF so is performant and synchronous (no surprises with files changing before you get to read them).

If you really want to do stuff with a file, you can try using the `system()` function to shell out to `cat`/`cp`/whatever. Bear in mind that the `system()` command will run asynchronously, so the `/proc/pid/environ` file might change or be deleted before you can read it.

**Reply by techneophyte** [reply]
Thanks @ajor. The `uid `included in the builtin variables is not the `realuserid `of the one who initiated the event. So, trying to find alternate options for this. I tried using system for cp as below inside my killsnoop bt file.

```
tracepoint:syscalls:sys_enter_kill
{
	@tpid[tid] = args.pid;
	@tsig[tid] = args.sig;
	system("cp /proc/1560334/cmdline /fileName");
} --unsafe
```
I am getting this error
```
killsnoop.bt:31:3-5: ERROR: syntax error, unexpected --, expecting {
} --unsafe
```
If I remove the unsafe part from the above tracepoint event, I see this error
```
killsnoop.bt:30:2-46: ERROR: system() is an unsafe function being used in safe mode
    system("cp /proc/1560334/cmdline /fileName");
```

Can you guide me here?

**Reply by ajor** [reply]
`uid` should return the "real UID of the task": https://github.com/torvalds/linux/blob/46d1907d1caaaaa422ae814c52065f243caa010a/include/linux/cred.h#L113

Is that not what you're after? bpftrace doesn't provide access to `suid`, `euid` or `fsuid` though.

**Reply by ajor** [reply]
For the issue with "--unsafe", you need to pass that as a command line argument to bpftrace. It doesn't work inside the script.

e.g.:
```
bpftrace --unsafe myscript.bt
```

**Reply by techneophyte** [reply]
@ajor, thanks for the suggestion. Running as unsafe from command line worked. Regarding the real uid, when I try to manually kill a process as sudo, I see the uid as root but not my uid. Any thoughts on how can I pull my user id using bpftrace? The environ file of the process that initiated this kill has my Id. That is the reason I am trying to get info from environ file. 

Also, using cat in bpftrace is working in safe mode but not cp from below snippet. Any thoughts on this?

```
tracepoint:syscalls:sys_enter_kill
{
	@tpid[tid] = args.pid;
	@tsig[tid] = args.sig;
	cat("/proc/%d/cmdline", @tpid[tid]);

	cp("/proc/15604/environ" "/newFile");

}
```
For cp, I am getting below error

```
ERROR: Unknown function: cp
       cp("/proc/1560334/environ /newFile");
```

**Reply by ajor** [reply]
Ah yeah, I see your issue - sudo will change the uid to root. From the `sudoers` man page it looks like there's an option to change this behaviour if that's something you want to do:
>      stay_setuid       Normally, when sudo executes a command the real and effective UIDs are set to
>                        the target user (root by default).  This option changes that behavior such that
>                        the real UID is left as the invoking user's UID.  In other words, this makes
>                        sudo act as a set-user-ID wrapper.

Otherwise, I think checking env vars like you're doing is the only way to go.


> Also, using cat in bpftrace is working in safe mode but not cp from below snippet. Any thoughts on this?

bpftrace doesn't let you execute external programs natively. `cat` works because it is a [builtin bpftrace function](https://github.com/bpftrace/bpftrace/blob/master/man/adoc/bpftrace.adoc#cat). bpftrace doesn't have a `cp` function - for that you'll need to shell out using `system("cp ...")`, which requires `--unsafe`.

**Reply by techneophyte** [reply]
@ajor Thanks for all your responses. I could not find much info on using `stay_setuid`. If possible, can you share any info on how to use it and what impact does it have on my system? 
Also, as using `system("cp ...")` is unsafe, can you share any info on how can I read the environ file inside bt script, store it in a variable and retrieve the user id property from it?
I can store the process ID from the bt script and read the environ file later. However, the main reason I want to do it within the bt script is that the proc environ file might be deleted by the time the other process reads it.

**Reply by ajor** [reply]
> I could not find much info on using `stay_setuid`. If possible, can you share any info on how to use it and what impact does it have on my system?

I think you'd just edit your `/etc/sudoers` file to add a line saying:
```
Defaults stay_setuid
```
You might be able to make it only apply to a single user with something like this, but I'm no expert:
```
Defaults:myusername stay_setuid
```

Tbh I'm not sure what kinds of impact it would have on a system.


> Also, as using `system("cp ...")` is unsafe, can you share any info on how can I read the environ file inside bt script, store it in a variable and retrieve the user id property from it? I can store the process ID from the bt script and read the environ file later.

This just isn't possible with BPF, sorry. The kernel doesn't provide any mechanism for reading files in BPF (or generally, anything that can block), which is why bpftrace implements the `cat()` function in userspace.

> However, the main reason I want to do it within the bt script is that the proc environ file might be deleted by the time the other process reads it.

Yep, that's a very valid concern with `system()` and `cat()`.

**Reply by techneophyte** [reply]
Thanks for all your help @ajor. One last question. Can we interfere with syscalls? Here, I am trying to get the information of how and why a process got terminated. If the process is being terminated in an unexpected way, can we leverage bpftrace to stop that event?

**Reply by ajor** [reply]
Yeah, it's possible to override the return values from certain kernel functions: https://github.com/bpftrace/bpftrace/blob/master/man/adoc/bpftrace.adoc#functions-override

Something like this script should do the job:
```
kprobe:__x64_sys_kill
{
  printf("kill from: %d, %s (%d)\n", uid, comm, pid);
  if (uid == 1234) {
    printf("blocking kill from user 1234\n");
    override(-1); // chose what return code you want here
  }
}
```

**Reply by techneophyte** [reply]
Thanks. Any thoughts if we can use inbuilt `cat `function to print output to a new file?

**Reply by ajor** [reply]
Sorry, I'm not sure if I'm misunderstanding but `cat()` can only display the contents of files, it can't write anything

**Reply by techneophyte** [reply]
Sorry for the confusion. What I mean here is, when I use `cat()`, the file is being printed on the terminal while bt script is executed. Is there a way to print it to a file instead of printing on the console?

**Reply by ajor** [reply]
Ah I see - there's no way to redirect just `cat`'s output. If you're not printing anything else then you can redirect all of bpftrace's output to a file with the `-o` command line flag, or just using shell redirects though.

**Reply by techneophyte** [reply]
Thanks. As part of my bpftrace script execution, I have to print multiple things. So, checking if there's a way to separate them

**Reply by ajor** [reply]
Just thought of a possible way: use bpftrace's JSON output (`-f json`). It'll tag events with their origin, e.g. "printf" or "cat".

**Reply by techneophyte** [reply]
Thank you so much for the suggestion. Let me try it to see if it fits my case.

**Reply by techneophyte** [reply]
@ajor for getting current process pid, I am trying to use `bpf_get_current_pid_tgid()` BPF helper inside a bt file. I am seeing an error
ERROR: Unknown function: bpf_get_current_pid_tgid

I also tried using creating a uprobe this way but got no luck

```
uprobe:executable:"function get_pid" {
   return bpf_get_current_pid_tgid() >> 32;
}

BEGIN
{
	printf("Current task PID: %d\n", get_pid());
	printf("Tracing kill() signals... Hit Ctrl-C to end.\n");    
	printf("%s %-15s %7s %5s %-16s %4s %6s %s\n", "loginuid+TPNAME", "TIME", "PID", "UID", "COMM", "SIG", "TPID", "RESULT");
}
```
Can you please help me on using these bpf helpers?

**Reply by ajor** [reply]
Responded in the other thread: https://github.com/bpftrace/bpftrace/discussions/3272
